File: src/render/native/src/status.cpp
#include "render_native/status.hpp"

namespace aura::render_native {

namespace {

int normalized_sample_count(const std::optional<int>& sample_count) {
    return sample_count.value_or(0);
}

}  // namespace

std::string format_initial_status(
    const std::optional<std::string>& db_path,
    const std::optional<int>& sample_count,
    const std::optional<std::string>& error
) {
    if (!db_path.has_value()) {
        return "Collecting telemetry...";
    }
    if (error.has_value()) {
        return "Collecting telemetry... | DVR unavailable: " + error.value();
    }
    return "Collecting telemetry... | DVR samples: " +
           std::to_string(normalized_sample_count(sample_count));
}

std::string format_stream_status(
    const std::optional<std::string>& db_path,
    const std::optional<int>& sample_count,
    const std::optional<std::string>& error
) {
    if (!db_path.has_value()) {
        return "Streaming telemetry";
    }
    if (error.has_value()) {
        return "Streaming telemetry | DVR unavailable: " + error.value();
    }
    return "Streaming telemetry | DVR samples: " +
           std::to_string(normalized_sample_count(sample_count));
}

}  // namespace aura::render_native

File: src/render/status.py
from __future__ import annotations

from typing import Protocol


class RecorderStatusProvider(Protocol):
    def get_status(self) -> tuple[str | None, int | None, str | None]:
        ...


def _normalized_sample_count(sample_count: int | None) -> int:
    return 0 if sample_count is None else sample_count


def format_initial_status(recorder: RecorderStatusProvider) -> str:
    db_path, sample_count, error = recorder.get_status()
    if db_path is None:
        return "Collecting telemetry..."
    if error is not None:
        return f"Collecting telemetry... | DVR unavailable: {error}"
    return (
        "Collecting telemetry... | "
        f"DVR samples: {_normalized_sample_count(sample_count)}"
    )


def format_stream_status(recorder: RecorderStatusProvider) -> str:
    db_path, sample_count, error = recorder.get_status()
    if db_path is None:
        return "Streaming telemetry"
    if error is not None:
        return f"Streaming telemetry | DVR unavailable: {error}"
    return f"Streaming telemetry | DVR samples: {_normalized_sample_count(sample_count)}"


__all__ = ["RecorderStatusProvider", "format_initial_status", "format_stream_status"]
